Support FP16 CpuAdam + Zero Stage 3

lz1oceani opened the pull request:
Add support for FP16 CpuAdam + Zero Stage 3

lz1oceani commented:
@microsoft-github-policy-service agree

tjruwase commented:
@lz1oceani, thanks for the PR. Can you please some description of the motivation and content of this PR? Thanks!

lz1oceani commented:
Motivation: Zero-stage3 requires all data to be stored in fp32 on the CPU, including weights, gradients + accumulated version, and optimization states. A 7B model needs 7 * 5 * 4 = 140GB of CPU memory, and this doubles for a 13B model. However, our cluster's memory is not often sufficient, as it typically runs additional jobs. To address this, I developed a fp16 version for Zero-stage3, providing robust optimization states, drawing inspiration from [this paper](https://arxiv.org/pdf/2310.18313.pdf).

Implementation: The original CPUAdam code partially supported fp16 weights but had bugs. I've added support for fp16 optimization states and corrected the fp16 weight support issue. For details, it is better the check the code.

I have check the code by fine-tuning Llama-7b. It can successfully converge.

tjruwase commented:
@lz1oceani, thanks. I have moved the description to the OP.

RezaYazdaniAminabadi commented on csrc/adam/cpu_adam_impl.cpp:
```diff
@@ -112,37 +148,65 @@ void Adam_Optimizer::Step_1(float* _params,
+        auto ret_remain = Step_1((_params + (rounded_size >> static_cast<int> (half_precision_param))),
+               (grads + (rounded_size >> static_cast<int> (half_precision_param))),
+               (_exp_avg + (rounded_size >> static_cast<int> (half_precision_opt_state))),
+               (_exp_avg_sq + (rounded_size >> static_cast<int> (half_precision_opt_state))),
```
can we please move this part of shifting to a separate utility function?

lz1oceani replied:
Yes. I can write a define for this.

RezaYazdaniAminabadi commented on csrc/includes/cpu_adam.h:
```diff
@@ -169,23 +198,45 @@ class Adam_Optimizer {
+    int OMP_MAX_THREADS = omp_get_max_threads();
+    float max_avg[OMP_MAX_THREADS], max_avg_sq[OMP_MAX_THREADS];
+    std::fill(max_avg, max_avg + OMP_MAX_THREADS, 0.0f);
+    std::fill(max_avg_sq, max_avg_sq + OMP_MAX_THREADS, 0.0f);
```
if I understand this part is related to Z3 and the serves as an initialization step, can we maybe move them to a different function and call it conditionally?

lz1oceani replied:
Let me think a little bit, it should be doable.
